scripts/check.unused.mjs

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env node
+
+import { readFileSync, readdirSync } from 'node:fs';
+import { basename, resolve } from 'node:path';
+
+const RED = '\x1b[31m';
+const GREEN = '\x1b[32m';
+const NC = '\x1b[0m'; // No Colour
+
+const DATA_DIR = 'src/frontend/src';
+const DATA_DIR_PATH = resolve(process.cwd(), DATA_DIR);
+
+const findSvelteFiles = (dir) =>
+	readdirSync(dir, { withFileTypes: true }).flatMap((file) => {
+		const res = resolve(dir, file.name);
+		return file.isDirectory()
+			? findSvelteFiles(res)
+			: file.isFile() && res.endsWith('.svelte')
+				? [res]
+				: [];
+	});
+
+const searchForFileName = ({ filename, dir }) =>
+	readdirSync(dir, { withFileTypes: true }).some((file) => {
+		const res = resolve(dir, file.name);
+		return file.isDirectory()
+			? searchForFileName({ filename, dir: res })
+			: file.isFile() && readFileSync(res, 'utf-8').includes(filename);
+	});
+
+const checkFileUsage = ({ filename, dir }) =>
+	filename.startsWith('+') || searchForFileName({ filename, dir });
+
+const logResult = (isUsed) => process.stdout.write(isUsed ? `${GREEN}.${NC}` : `${RED}x${NC}`);
+
+const main = async () => {
+	console.log(`${NC}Scanning ${DATA_DIR} folder to find all .svelte files`);
+	console.log(`  ${GREEN}.${NC} means the .svelte is imported in another file`);
+	console.log(`  ${RED}x${NC} means the .svelte is not imported and should likely be removed`);
+
+	const svelteFiles = findSvelteFiles(DATA_DIR_PATH);
+	const unusedFiles = svelteFiles.filter((file) => {
```

Comment:
I'm not sure what the best option is for performance reasons. If I understand correctly, we are reading the content of all files for every single file.

For example, we find a.svelte and then we check every file to see if a.svelte is present.

Would it be more efficient to build a list of all filenames and then visit each file to read its content and eliminate the entries until there are no more entries to eliminate or all files have been browsed? If ultimately the list is not empty, those remaining entries are the leaves.

Just thinking that if we add the script to the CI if would be nice if it runs the fastest.

Reply:
you mean store the content of each file, right? We would visit them just once and store the content (separately or all together). I think it would be just a memory "issue" at that point, no?

Reply:
No I mean something like pseud code:

```
const listOfAllSvelteFiles = findSvelteFiles(); // [src/yolo.svelte, src/+page.svelte]

cont listOfPotentialUnusedFiles = [...listOfAllSvelteFiles]);

for (const file of listOfAllSvelteFiles) {
   const content = readFile(file);
   
   const usedFiles = listOfPotentialUnusedFiles.filter(file => content.include(file));
   
   listOfPotentialUnusedFiles = [...listOfPotentialUnusedFiles.filter(file => !usedFiles.includes(file))];
   
   if (listOfPotentialUnusedFiles.length === 0) {
      process.exit(0); // No used files
  }
}

listOfPotentialUnusedFiles.forEach(file => console.log("Unused file:", file));
```

While we have to iterate over an array, I was just thinking that that way, even if naive, maybe there is a chance that we would have to read less content and therefore less file system access which potentially is slower than filtering the array?

Just an idea of course, that might be less performant.

Reply:
Ah! Right! Now I understand! It makes sense indeed! Check now please